File: 362_design_hit_counter.java
class HitCounter {
    public static final int interval = 300;
    int currTime;
    int numEles;
    Deque<Integer> queue;
    public HitCounter() {
        currTime = Integer.MIN_VALUE;
        numEles = 0;
        queue = new LinkedList<>();
    }

    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    public void hit(int timestamp) {
        queue.offerLast(timestamp);
        numEles++;
        currTime = timestamp;
    }

    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    public int getHits(int timestamp) {
        while (!queue.isEmpty()) {
            int val = queue.peekFirst();
            if (timestamp - val >= interval) {
                queue.pollFirst();
                numEles--;
            } else {
                break;
            }
        }
        return this.numEles;
    }
}

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter obj = new HitCounter();
 * obj.hit(timestamp);
 * int param_2 = obj.getHits(timestamp);
 */

File: python/362_design_hit_counter.py
class HitCounter:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.count_map = collections.defaultdict(int)

    def update(self, timestamp: int) -> None:
        deleted = []
        for t, hits in self.count_map.items():
            # discard larger than 5 mins
            if timestamp - t >= 300:
                deleted.append(t)
        for key in deleted: del self.count_map[key]

    def hit(self, timestamp: int) -> None:
        """
        Record a hit.
        @param timestamp - The current timestamp (in seconds granularity).
        """
        self.update(timestamp)
        self.count_map[timestamp] += 1

    def getHits(self, timestamp: int) -> int:
        """
        Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity).
        """
        self.update(timestamp)
        total = 0
        for val in self.count_map.values():
            total += val
        return total


# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)
